pages/agent/v3/signed-pipelines.md

```diff
@@ -0,0 +1,122 @@
+
+On instances that will verify jobs, add:
+```ini
+job-verification-jwks-path=<path to verification keys>
+```
+
+### 3. Signing the pipeline's stored configuration
+
+[ed: this section feels clunky to me, and i'm not sure the best way to explain it - we need to sign the initial steps of a pipeline as well as the steps that get uploaded by an agent, but it's difficult to concisely convey all the concepts involved]
+
+Builds in Buildkite pipelines are initialized using steps that are configured on the pipeline itself. In the course of running a job, the Buildkite Agent can [upload further steps to the pipeline](link to some info about this elsewhere in the docs). We will refer to the former as "configured steps" and the latter as "uploaded steps". It's common for the configured steps to be a single step that uploads the remainder of pipeline from [a YAML file in the repository](link to info about ./buildkite.yaml).
+
+> Some older Buildkite pipelines don't use YAML for their stored pipeline configuration - these can't be signed, and will have to be migrated to use the new format to be used with signed pipelines - see [Migrating to YAML steps](/docs/tutorials/pipeline_upgrade)
+
+You may have noticed that when we configured the signing instances, we gave them the capability to both sign and verify jobs. For maximum security, the configured steps - usually a single step that uploads (and signs!) the remainder of the pipeline - should themselves be signed.
+
+The solution here is to add static signatures to the steps in your pipeline's stored configuration.
+
+To do this, copy the YAML from the editor on the pipeline's page in Buildkite, and paste it into a file on your machine. Then, run:
+```bash
+buildkite-agent pipeline upload --jwks-file-path <path to private keys> --signing-key-id <key id> --dry-run <path to pipeline file>
+```
+
+for example, if you had a file called `pipeline.yml` in the current directory, and you wanted to sign it with the key we generated earlier, you'd run:
+```bash
+buildkite-agent pipeline upload --jwks-file-path /path/to/private/key.json --signing-key-id my-key-id --dry-run pipeline.yml
+```
+
+This will output the pipeline YAML, with the signatures added to the steps. You can then copy this YAML back into the editor on the pipeline's page in Buildkite, and save it.
+
+## Rotating signing keys
+
+Because signed pipelines use JWKS as their key format, rotating keys is easy. To rotate your keys, simply generate a new key pair, and add the new keys to your existing key sets, being careful not to mix public and private keys. Once the updated key sets are in place, you can update the `signing-key-id` on your signing agents to use the new key ID. Verifying agents will automatically the public key with the matching key ID, if it's present.
```

Comment:
>Verifying agents will automatically [use] the public key with the matching key ID, if it's present.

Isn't it signing agents that need the key ID? and verifying agents just try all the keys to see if one works? Or is the key ID now in the signature itself, so the verifying agent can reference that?

Reply:
the key ID is encoded in the header of the signature - here's one i pulled from the agent pipeline: `eyJhbGciOiJFUzI1NiIsImtpZCI6IjEifQ..Ue0tFd7jaZj66huix2TZ1UanAB4vHlWZn1utvXQDjSMmJDQ1wWVG8fDCdboZrUcnxzNen1MaU8NfX_vOr5T7sQ`

when you base64 decode the first segment there, it gives: `{"alg":"ES256","kid":"1"}` - our key id in the agent is imaginatively called `1`. Our library then pulls this out of the header and looks for a key with that key ID in the verification JWKS.

if we didn't include a key ID (which isn't allowed as is), the library would then try each key in sequence.

Reply:
Gotcha - thanks! In that case:
```suggestion
Because signed pipelines use JWKS as their key format, rotating keys is easy. To rotate your keys, simply generate a new key pair, and add the new keys to your existing key sets, being careful not to mix public and private keys. Once the updated key sets are in place, you can update the `signing-key-id` on your signing agents to use the new key ID. Verifying agents will automatically use the public key with the matching key ID, if it's present.
```